File: src/simulations/NewtonianGravity/Vector.java
package simulations.NewtonianGravity;

/**
 * Vector class - represents a mathematical vector as an array.
 * Supports common vector operations needed for physics simulations.
 */
public class Vector {
    private double[] data;
    
    /**
     * Creates a vector from an array of components.
     */
    public Vector(double[] components) {
        this.data = components.clone();
    }
    
    /**
     * Creates a vector with specified number of dimensions, initialized to zero.
     */
    public Vector(int dimensions) {
        this.data = new double[dimensions];
    }
    
    
    /**
     * Gets the number of dimensions.
     */
    public int dimensions() {
        return data.length;
    }
    
    /**
     * Gets a component by index.
     */
    public double get(int index) {
        return data[index];
    }
    
    /**
     * Sets a component by index.
     */
    public void set(int index, double value) {
        data[index] = value;
    }
    
    /**
     * Gets the underlying array (defensive copy).
     */
    public double[] getData() {
        return data.clone();
    }
    
    /**
     * Adds another vector to this vector (element-wise).
     */
    public Vector add(Vector other) {
        if (this.data.length != other.data.length) {
            throw new IllegalArgumentException("Vectors must have same dimensions");
        }
        double[] result = new double[data.length];
        for (int i = 0; i < data.length; i++) {
            result[i] = this.data[i] + other.data[i];
        }
        return new Vector(result);
    }
    
    /**
     * Subtracts another vector from this vector.
     */
    public Vector subtract(Vector other) {
        if (this.data.length != other.data.length) {
            throw new IllegalArgumentException("Vectors must have same dimensions");
        }
        double[] result = new double[data.length];
        for (int i = 0; i < data.length; i++) {
            result[i] = this.data[i] - other.data[i];
        }
        return new Vector(result);
    }
    
    /**
     * Multiplies this vector by a scalar.
     */
    public Vector multiply(double scalar) {
        double[] result = new double[data.length];
        for (int i = 0; i < data.length; i++) {
            result[i] = this.data[i] * scalar;
        }
        return new Vector(result);
    }


    public Vector divide(double scalar) {
        return multiply(1.0 / scalar);
    }
    /**
     * Computes the dot product with another vector.
     */
    public double dot(Vector other) {
        if (this.data.length != other.data.length) {
            throw new IllegalArgumentException("Vectors must have same dimensions");
        }
        double result = 0.0;
        for (int i = 0; i < data.length; i++) {
            result += this.data[i] * other.data[i];
        }
        return result;
    }

    public Vector cross(Vector other) {
        if (this.data.length != 3 || other.data.length != 3) {
            throw new IllegalArgumentException("Vectors must be 3D to take the cross product");
        }
        double[] result = new double[3];
        result[0] = this.data[1] * other.data[2] - this.data[2] * other.data[1];
        result[1] = this.data[2] * other.data[0] - this.data[0] * other.data[2];
        result[2] = this.data[0] * other.data[1] - this.data[1] * other.data[0];
        return new Vector(result);
    }
    
    /**
     * Computes the magnitude (length) of the vector.
     */
    public double magnitude() {
        double sumSquares = 0.0;
        for (double d : data) {
            sumSquares += d * d;
        }
        return Math.sqrt(sumSquares);
    }
    
    /**
     * Returns a normalized (unit) vector.
     */
    public Vector normalize() {
        double mag = magnitude();
        if (mag == 0.0) {
            throw new ArithmeticException("Cannot normalize zero vector");
        }
        return multiply(1.0 / mag);
    }
    
    /**
     * Creates a copy of this vector.
     */
    public Vector clone() {
        return new Vector(this.data);
    }
    
    /**
     * Returns a string representation.
     */
    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder("(");
        for (int i = 0; i < data.length; i++) {
            if (i > 0) sb.append(", ");
            sb.append(data[i]);
        }
        sb.append(")");
        return sb.toString();
    }
}

File: resources/web/gravity3d/physics/Vector.js
/**
 * Vector class - represents a mathematical vector as an array.
 * Supports common vector operations needed for physics simulations.
 */
class Vector {
    constructor(components) {
        if (Array.isArray(components)) {
            this.data = [...components];
        } else if (typeof components === 'number') {
            // Create zero vector with specified dimensions
            this.data = new Array(components).fill(0);
        } else {
            throw new Error('Vector must be initialized with array or dimension count');
        }
    }
    
    /**
     * Gets the number of dimensions.
     */
    dimensions() {
        return this.data.length;
    }
    
    /**
     * Gets a component by index.
     */
    get(index) {
        return this.data[index];
    }
    
    /**
     * Sets a component by index.
     */
    set(index, value) {
        this.data[index] = value;
    }
    
    /**
     * Gets the underlying array (defensive copy).
     */
    getData() {
        return [...this.data];
    }
    
    /**
     * Adds another vector to this vector (element-wise).
     */
    add(other) {
        if (this.data.length !== other.data.length) {
            throw new Error('Vectors must have same dimensions');
        }
        const result = new Array(this.data.length);
        for (let i = 0; i < this.data.length; i++) {
            result[i] = this.data[i] + other.data[i];
        }
        return new Vector(result);
    }
    
    /**
     * Subtracts another vector from this vector.
     */
    subtract(other) {
        if (this.data.length !== other.data.length) {
            throw new Error('Vectors must have same dimensions');
        }
        const result = new Array(this.data.length);
        for (let i = 0; i < this.data.length; i++) {
            result[i] = this.data[i] - other.data[i];
        }
        return new Vector(result);
    }
    
    /**
     * Multiplies this vector by a scalar.
     */
    multiply(scalar) {
        const result = new Array(this.data.length);
        for (let i = 0; i < this.data.length; i++) {
            result[i] = this.data[i] * scalar;
        }
        return new Vector(result);
    }
    
    /**
     * Divides this vector by a scalar.
     */
    divide(scalar) {
        return this.multiply(1.0 / scalar);
    }
    
    /**
     * Computes the dot product with another vector.
     */
    dot(other) {
        if (this.data.length !== other.data.length) {
            throw new Error('Vectors must have same dimensions');
        }
        let result = 0.0;
        for (let i = 0; i < this.data.length; i++) {
            result += this.data[i] * other.data[i];
        }
        return result;
    }
    
    /**
     * Computes the cross product with another vector (3D only).
     */
    cross(other) {
        if (this.data.length !== 3 || other.data.length !== 3) {
            throw new Error('Vectors must be 3D to take the cross product');
        }
        return new Vector([
            this.data[1] * other.data[2] - this.data[2] * other.data[1],
            this.data[2] * other.data[0] - this.data[0] * other.data[2],
            this.data[0] * other.data[1] - this.data[1] * other.data[0]
        ]);
    }
    
    /**
     * Computes the magnitude (length) of the vector.
     */
    magnitude() {
        let sumSquares = 0.0;
        for (let d of this.data) {
            sumSquares += d * d;
        }
        return Math.sqrt(sumSquares);
    }
    
    /**
     * Returns a normalized (unit) vector.
     */
    normalize() {
        const mag = this.magnitude();
        if (mag === 0.0) {
            throw new Error('Cannot normalize zero vector');
        }
        return this.multiply(1.0 / mag);
    }
    
    /**
     * Adds another vector to this vector in-place.
     */
    addTo(other) {
        if (this.data.length !== other.data.length) {
            throw new Error('Vectors must have same dimensions');
        }
        for (let i = 0; i < this.data.length; i++) {
            this.data[i] += other.data[i];
        }
        return this;
    }
    
    /**
     * Creates a copy of this vector.
     */
    clone() {
        return new Vector(this.data);
    }
    
    /**
     * Converts to Three.js Vector3 (if 3D).
     */
    toThreeVector3() {
        if (this.data.length < 3) {
            throw new Error('Vector must be at least 3D to convert to Vector3');
        }
        return new THREE.Vector3(this.data[0], this.data[1], this.data[2]);
    }
    
    /**
     * Returns a string representation.
     */
    toString() {
        return '(' + this.data.join(', ') + ')';
    }
}
